docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,179 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
+
+**Property 3**: Messages from interleaving calls have no reliable execution ordering.
+
+Let's consider the above example code again, and assume the method `example` is called twice in parallel, the resulting calls being Call 1 and Call 2. The following illustration shows two possible message orderings. On the left, the first call's messages are scheduled first, and only then the second call's messages are executed. On the right, we see another possible message scheduling, where the first messages of each call are executed first. Your code should result in a correct state regardless of the message ordering.
+
+![example_orderings](_attachments/example_orderings.png)
+
+**Property 4**: On a trap / panic, modifications to the canister state for the current message are not applied.
+
+For example, if a trap in the second message (blue box) of the above example occurs, canister state changes resulting from that message, even earlier in the blue box, are discarded. However, note that any state changes from earlier messages and in particular the first message (orange box) have been applied, as that message executed successfully. 
+
+For more details, refer to the Interface Specification [section on ordering guarantees](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ordering_guarantees) and the section on [abstract behaviour](https://internetcomputer.org/docs/current/references/ic-interface-spec/#abstract-behavior) which defines message execution in more detail. 
+
+### Avoid traps after await
 
 #### Security Concern
 
-Panics and traps roll back the canister state. So any state change followed by a trap or panic is of concern. This is also an important concern when inter-canister calls are made. If a panic/trap occurs after an `await` to an inter-canister call, then the state is reverted to the snapshot before the inter-canister call callback invocation (and not before the entire call!).
+Traps / panics roll back the canister state, as described in Property 4 above. So any state change followed by a trap or panic can be risky. This is also an important concern when inter-canister calls are made. If a panic/trap occurs after an `await` to an inter-canister call, then the state is reverted to the snapshot before the inter-canister call callback invocation, and not before the entire call!.
 
 This may e.g. lead to the following issues:
 
--   If state changes before an inter-canister call leave the state inconsistent and there is a panic after the inter-canister call, this results in inconsistent canister state.
+- Suppose some state changes are applied and then an inter-canister call is issued. Also, assume that these state changes leave the canister in an inconsistent state, and that state is only made consistent again in the callback. Now if there is a trap in the callback, this leaves the canister in an inconsistent state. 
```

Comment:
It'd be great if we could give a short concrete example. I could cook something up, but maybe you have a concrete one from your reviews?

Reply:
Yes that makes sense! Added this concrete example:
* _A concrete bug of this kind is the following. Assume an inter-canister call is issued to transfer funds. In the callback, the canister accounts for having made that transfer by reflecting that fact the canister storage. However, suppose the callback also updates some usage statistics data, which eventually leads to a trap when some data structure becomes full. As soon as that is the case, the canister ends up in an inconsistent state because the state changes in the callback are no longer applied and thus the transfers are not correctly accounted for._